twilight-gateway/src/json.rs

```diff
@@ -8,96 +8,48 @@ pub use simd_json::to_string;
-use std::{
-    error::Error,
-    fmt::{Debug, Display, Formatter, Result as FmtResult},
-};
 use twilight_model::gateway::{
-    event::{GatewayEvent, GatewayEventDeserializer},
+    event::{Event, GatewayEventDeserializer},
     OpCode,
 };
 
-/// Error occurred due to an unknown event and opcode pair.
-#[derive(Debug)]
-#[non_exhaustive]
-pub(crate) struct UnknownEventError {
-    /// Event type in the payload.
-    pub event_type: Option<String>,
-    /// Opcode in the payload.
-    pub opcode: Option<u8>,
-}
-
-impl Display for UnknownEventError {
-    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
-        f.write_str("unknown opcode/dispatch event type: ")?;
-        Debug::fmt(&self.opcode, f)?;
-        f.write_str("/")?;
-
-        Debug::fmt(&self.event_type, f)
-    }
-}
-
-impl Error for UnknownEventError {}
-
-/// Parse a JSON encoded event into a gateway event if its type is in
-/// `wanted_event_types`.
+/// Deserialize a websocket message into an event if `wanted_event_types`
+/// contains its type.
 ///
-/// This function can be used together with [`Shard::next_message`] for greater
-/// control of the deserialization process without giving up the ease of use of
-/// [`Shard::next_event`].
-///
-/// Returns [`None`] if the event type is not contained inside of
-/// `wanted_event_types`.
+/// Close messages are always considered wanted and map onto
+/// [`Event::GatewayClose`].
 ///
 /// # Errors
 ///
-/// Returns a [`ReceiveMessageErrorType::Deserializing`] error if the event
-/// could not be deserialized.
-///
-/// [`Shard::next_event`]: crate::Shard::next_event
-/// [`Shard::next_message`]: crate::Shard::next_message
-pub fn parse(
-    event: String,
+/// Returns a [`ReceiveMessageErrorType::Deserializing`] error if the *known*
+/// event could not be deserialized.
+pub fn deserialize_wanted(
+    message: Message,
     wanted_event_types: EventTypeFlags,
-) -> Result<Option<GatewayEvent>, ReceiveMessageError> {
+) -> Result<Option<Event>, ReceiveMessageError> {
+    let event = match message {
+        Message::Close(frame) => return Ok(Some(Event::GatewayClose(frame))),
+        Message::Text(event) => event,
+    };
+
     let Some(gateway_deserializer) = GatewayEventDeserializer::from_json(&event) else {
         return Err(ReceiveMessageError {
             kind: ReceiveMessageErrorType::Deserializing { event },
-            source: Some(Box::new(UnknownEventError {
-                event_type: None,
-                opcode: None,
-            })),
+            source: None,
         });
     };
 
     let Some(opcode) = OpCode::from(gateway_deserializer.op()) else {
-        let opcode = gateway_deserializer.op();
-
-        return Err(ReceiveMessageError {
-            kind: ReceiveMessageErrorType::Deserializing { event },
-            source: Some(Box::new(UnknownEventError {
-                event_type: None,
-                opcode: Some(opcode),
-            })),
-        });
+        return Ok(None);
```

Comment:
I'm not sure I agree that unknown opcodes should not error. This indicates a rather serious protocol change on Discord's end

Reply:
We already catch and log unknown opcodes in `Shard::process` and Discord would never break their API by introducing a new, critical, opcode. If it were to blow up, returning an error here wouldn't mitigate anything

Reply:
Having it in `Shard::process` seems good enough.